### ISSUE TITLE
`mix release` produces start.script which starts dependencies of a load-only dependency

### ISSUE THREAD

**Issue by jjcarstens** [elixir-lang/elixir#11815]
### Environment

* Elixir & Erlang/OTP versions (elixir --version): **Elixir 1.13.4 (compiled with Erlang/OTP 24)**
* Operating system: **MacOS 12.3**

### Current behavior

TL;DR - if you set a dependency `{:b, "~> 0.1", runtime: false}` and then load it in the release with `applications: [b: :load]`, the dependents of `:b` will still be started even though `:b` itself is not (assuming the dependents only belong to `:b`. If `:a` and `:b` share dependents, then I expect those deps to start with `:a` like they currently do)

Example setup can be found at: https://github.com/jjcarstens/release_load_deps

<details>
<summary>Explanation & Examples</summary>

I have an application setup where `:a` depends on `:b` and `:b` depends on `:json`:
```
> a √ % mix deps.tree
a
└── b (../b)
    └── json ~> 1.0 (Hex package)
```

When I make a release of `:a`, I only want `:b` to be loaded so that its available in the release and I can start it later on at Runtime at my discretion. The mix.exs of `:a` would look something like this:

```elixir
def project do
  [
    app: :a,
    deps: [{:b, path: "../b", runtime: false]},
    releases: [{:a, [applications: [b: :load]]}]
  ]
end
```

And `:b` has only one dep in ` mix.exs`:
```elixir
def project do
  [
    app: :b,
    deps: [{:json, "~> 1.0"}]
  ]
end
```

Through transitive property, `:a` now also has a dependency on `:json`. However, since `:json` is not a dependency directly of `:a` or another other of dependencies of `:a` (if there were any in this example), I would expect that `:json` would only be loaded as well since its parent app `:b` was only loaded.

It seems this is not the case if you were to run `mix release` and run app `:a`:
```elixir
> a √ % _build/dev/rel/a/bin/a start_iex 
Erlang/OTP 24 [erts-12.3] [source] [64-bit] [smp:10:10] [ds:10:10:10] [async-threads:1]

Interactive Elixir (1.13.4) - press Ctrl+C to exit (type h() ENTER for help)
iex(a@megakopf)1> Application.started_applications
[
  {:iex, 'iex', '1.13.4'},
  {:a, 'a', '0.1.0'},
  {:json, 'The First Native Elixir library for JSON encoding and decoding', '1.4.1'},
  {:logger, 'logger', '1.13.4'},
  {:sasl, 'SASL  CXC 138 11', '4.1.2'},
  {:elixir, 'elixir', '1.13.4'},
  {:compiler, 'ERTS  CXC 138 10', '8.1'},
  {:stdlib, 'ERTS  CXC 138 10', '3.17.1'},
  {:kernel, 'ERTS  CXC 138 10', '8.3.1'}
]
iex(a@megakopf)2> 
```

And also confirmed in the generated `start.script`:
```
     ...
     {progress,applications_loaded},
     {apply,{application,start_boot,[kernel,permanent]}},
     {apply,{application,start_boot,[stdlib,permanent]}},
     {apply,{application,start_boot,[compiler,permanent]}},
     {apply,{application,start_boot,[elixir,permanent]}},
     {apply,{application,start_boot,[sasl,permanent]}},
     {apply,{application,start_boot,[logger,permanent]}},
     {apply,{application,start_boot,[json,permanent]}},
     {apply,{application,start_boot,[a,permanent]}},
     {progress,started}]}.
```
</details>

### Expected behavior

If the deps of a loaded dep are not shared with the parent, I would expect them not to be started. This is something discovered while messing around with [shoehorn](https://github.com/nerves-project/shoehorn) for Nerves devices and we were originally thinking about adjusting things there if this is totally by design in Elixir releases. If not, then we can contribute it back to the Elixir core.

/cc @fhunleth

**Comment by josevalim** [comment]
Excellent report!

**Comment by josevalim** [comment]
Hi @jjcarstens, I have been looking into this and I am wondering how far should those configurations cascade. For example, if I specify an app is temporary, should all of its children be marked as temporary too? Does this also apply to none or loaded?

@tsloughter, do you have any thoughts here? Is this also how relx behaves?

**Comment by josevalim** [comment]
Closing in favor of the PR! But still waiting on @tsloughter or more external feedback before merging. :)

**Comment by jjcarstens** [comment]
Sorry, been out for a few days

I've breezed over the PR and looks like this discussion has kicked off there. But just to confirm, I would say "yes, config should traverse to all children unless it is a child of another dep as well with a permanent load"

Thanks!

**Comment by tsloughter** [comment]
rebar3/relx does not set children to be temporary.

**Comment by josevalim** [comment]
@tsloughter so if i set the parent to temporary or load, it does not automatically cascade to children?

**Comment by tsloughter** [comment]
@josevalim right. Each one must be set manually if they need to also be temporary or load.
